frontend/src/lib/hooks/useScrollSpy.ts

```diff
@@ -0,0 +1,126 @@
+/**
+ * Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { useLayoutEffect, useMemo, useRef, useCallback } from "react"
+
+/**
+ * Creates a debounced function that delays invoking `fn` until after `ms`
+ * milliseconds have passed since the last time the debounced function was
+ * invoked.
+ *
+ * @param fn - A function to debounce. `fn` is called after the debounced
+ * function has not been called for `ms` milliseconds.
+ *
+ * @param ms - The delay in milliseconds before `fn` is executed.
+ *
+ * The debounced function behaves as follows:
+ * - It will be invoked immediately if 0 is passed for `ms`.
+ * - It will be invoked immediately for the first call in any case.
+ * - For subsequent calls, if less than `ms` milliseconds have passed since
+ * the last invocation, a new invocation of `fn` is scheduled for `ms`
+ * milliseconds after the last invocation.
+ * - If it is invoked and `ms` milliseconds have passed since the last
+ * invocation, `fn` is executed immediately and the timestamp is updated.
+ * - If a new invocation of the debounced function is scheduled and it
+ * is invoked again before the scheduled invocation, the scheduled
+ * invocation is canceled and a new one is scheduled `ms` milliseconds
+ * after the latest invocation.
+
+ *
+ * @returns A debounced version of the `fn` function.
+ */
+export function debounce(
```

Comment:
nit: Could this be somehow merged with our other `debounce` function here? 

https://github.com/streamlit/streamlit/blob/3407ec723cd6c12c085907bf743c1a27e96e3ac2/frontend/src/lib/util/utils.ts#L32